controller/src/main/java/io/pravega/controller/server/retention/AutoRetentionBucketService.java

```diff
@@ -0,0 +1,163 @@
+                    notificationLoop = Futures.loop(() -> !stop.get(), this::processNotification, executor);
+                })
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                });
+    }
+
+    private CompletableFuture<Void> processNotification() {
+        return CompletableFuture.runAsync(() -> {
+            StreamNotification notification = Exceptions.handleInterrupted(notifications::take);
+
+            StreamImpl stream = !notification.getType().equals(StreamNotification.NotificationType.ConnectivityError) ?
+                    new StreamImpl(notification.getScope(), notification.getStream()) : null;
+
+            switch (notification.getType()) {
+                case StreamAdded:
+                    retentionFutureMap.putIfAbsent(stream, getStreamRetentionFuture(stream));
+                    break;
+                case StreamRemoved:
+                    retentionFutureMap.remove(stream).cancel(true);
+                    break;
+                case StreamUpdated:
+                    // if retention has been disabled then cancel the future.
+                    // if retention has been enabled then add the future if it does not exist.
+                    // For now we will do nothing.
+                    break;
+                case ConnectivityError:
+                    break;
+            }
+        }, executor);
+    }
+
+    private CompletableFuture<Void> getStreamRetentionFuture(StreamImpl stream) {
+        return Futures.loopWithDelay(() -> !stop.get(), () -> performRetention(stream),
+                Duration.ofMinutes(Config.MINIMUM_RETENTION_FREQUENCY_IN_MINUTES).toMillis(), executor);
+    }
+
+    private CompletableFuture<Void> performRetention(StreamImpl stream) {
+        OperationContext context = streamMetadataStore.createContext(stream.getScope(), stream.getStreamName());
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getConfiguration(stream.getScope(), stream.getStreamName(), context, executor)
+                .thenCompose(config -> streamMetadataTasks.autoRetention(stream.getScope(), stream.getStreamName(),
+                        config.getRetentionPolicy(), System.currentTimeMillis(), context))
+                .exceptionally(e -> {
+                    log.warn("Exception thrown while performing auto retention for stream {}, {}", stream, e);
+                    throw new CompletionException(e);
+                }), RetryHelper.UNCONDITIONAL_PREDICATE, 5, executor)
+                .exceptionally(e -> {
+                    log.warn("Unable to perform retention for stream {}, {}. Ignoring, retention will be attempted in next cycle.", stream, e);
+                    return null;
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        this.stop.set(true);
```

Comment:
well i was using stop because we started the loop before doing notifyStarted which meant state may not be running. 
changing that to start loop after notifyStarted